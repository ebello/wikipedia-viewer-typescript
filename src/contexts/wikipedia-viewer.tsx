import React, { useState, useEffect, ReactNode } from 'react';
import PropTypes from 'prop-types';

interface ViewingHistoryItem {
  timestamp: number,
  title: string,
}

interface WikipediaContextValue {
  readonly viewingHistory: ViewingHistoryItem[],
  addToViewingHistory: (title: string) => void,
  clearViewingHistory: () => void,
}

type ContextProps = {
  children: ReactNode,
};

const WikipediaViewerContext = React.createContext<WikipediaContextValue>({
  viewingHistory: [],
  addToViewingHistory: () => null,
  clearViewingHistory: () => null,
});

const WikipediaViewer = ({ children }: ContextProps): JSX.Element => {
  const viewingHistoryStr = window?.localStorage.getItem('viewingHistory');
  const defaultHistory = JSON.parse(viewingHistoryStr || '') as ViewingHistoryItem[] || [];
  const [viewingHistory, setViewingHistory] = useState(defaultHistory);

  const addToViewingHistory = (title: string) => {
    setViewingHistory([...viewingHistory, {
      timestamp: Date.now(),
      title,
    }]);
  };

  const clearViewingHistory = () => setViewingHistory([]);

  useEffect(() => {
    window.localStorage.setItem('viewingHistory', JSON.stringify(viewingHistory));
  }, [viewingHistory]);

  return (
    <WikipediaViewerContext.Provider
      value={{
        viewingHistory,
        addToViewingHistory,
        clearViewingHistory,
      }}
    >
      {children}
    </WikipediaViewerContext.Provider>
  );
};

WikipediaViewer.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
};

export default WikipediaViewer;

export { WikipediaViewerContext };
