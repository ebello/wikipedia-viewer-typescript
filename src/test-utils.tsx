import React, { ReactNode } from 'react';
import { render, RenderResult, RenderOptions } from '@testing-library/react';
import {
  createHistory,
  createMemorySource,
  LocationProvider,
  History,
} from '@reach/router';
import { ThemeProvider } from 'theme-ui';
import theme from './theme';

// const AllTheProviders = ({ children }: { children: ReactNode }): React.FC<{ children: ReactNode }> => (
//   <ThemeProvider theme={theme}>
//     {children}
//   </ThemeProvider>
// );

type ProviderProps = {
  children: ReactNode,
};

const AllTheProviders: React.FunctionComponent<ProviderProps> = ({ children }: ProviderProps) => (
  <ThemeProvider theme={theme}>
    {children}
  </ThemeProvider>
);

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
): RenderResult => render(ui, { wrapper: AllTheProviders, ...options });

export const renderWithRouter = (
  ui: React.ReactElement,
  { route = '/', history = createHistory(createMemorySource(route)) } = {},
): RenderResult & { history: History } => ({
  ...render(<LocationProvider history={history}>{ui}</LocationProvider>),
  // adding `history` to the returned utilities to allow us
  // to reference it in our tests (just try to avoid using
  // this to test implementation details).
  history,
});

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };

export const sleep = (ms: number): Promise<unknown> => new Promise(
  (resolve) => setTimeout(resolve, ms),
);
