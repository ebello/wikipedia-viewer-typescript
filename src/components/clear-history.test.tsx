import React from 'react';
import { render, screen, fireEvent } from '../test-utils';
// import { render, screen } from '@testing-library/react';
import { WikipediaViewerContext } from '../contexts/wikipedia-viewer';
import ClearHistory from './clear-history';

describe('Clear History', () => {
  it('renders null with no context', () => {
    const { container } = render(<ClearHistory />);
    expect(container).toBeEmptyDOMElement();
  });

  it('renders null with no viewingHistory', () => {
    const { container } = render(
      <WikipediaViewerContext.Provider
        value={{
          viewingHistory: [],
          addToViewingHistory: () => {},
          clearViewingHistory: () => {},
        }}
      >
        <ClearHistory />
      </WikipediaViewerContext.Provider>,
    );
    expect(container).toBeEmptyDOMElement();
  });

  it('renders with viewingHistory', () => {
    const { asFragment } = render(
      <WikipediaViewerContext.Provider
        value={{
          viewingHistory: [{ timestamp: 12345, title: 'sample title' }],
          addToViewingHistory: () => {},
          clearViewingHistory: () => {},
        }}
      >
        <ClearHistory />
      </WikipediaViewerContext.Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('calls clear viewing history on click', () => {
    const clearViewingHistory = jest.fn();
    render(
      <WikipediaViewerContext.Provider
        value={{
          viewingHistory: [{ timestamp: 12345, title: 'sample title' }],
          addToViewingHistory: () => {},
          clearViewingHistory,
        }}
      >
        <ClearHistory />
      </WikipediaViewerContext.Provider>,
    );
    fireEvent.click(screen.getByText('Clear history'));
    expect(clearViewingHistory).toHaveBeenCalled();
  });

  it('calls clear viewing history on key up', () => {
    const clearViewingHistory = jest.fn();
    render(
      <WikipediaViewerContext.Provider
        value={{
          viewingHistory: [{ timestamp: 12345, title: 'sample title' }],
          addToViewingHistory: () => {},
          clearViewingHistory,
        }}
      >
        <ClearHistory />
      </WikipediaViewerContext.Provider>,
    );
    fireEvent.keyUp(screen.getByText('Clear history'));
    expect(clearViewingHistory).toHaveBeenCalled();
  });
});
