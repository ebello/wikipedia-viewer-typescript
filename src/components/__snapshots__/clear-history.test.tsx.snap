// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clear History renders with viewingHistory 1`] = `
<DocumentFragment>
  .emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: inline-block;
  text-align: center;
  line-height: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: inherit;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 8px;
  padding-bottom: 8px;
  color: white;
  background-color: var(--theme-ui-colors-primary,#f00);
  border: 0;
  border-radius: 4px;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 8px;
  padding-right: 8px;
  font-size: 85%;
  margin-left: 8px;
  background-color: var(--theme-ui-colors-bgmuted,#b3c1ca);
}

<button
    class="emotion-0"
  >
    Clear history
  </button>
</DocumentFragment>
`;
